File: internal/driver/nats/clients.go
package nats

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/amigoer/mq-studio/internal/model"
)

// Attribute keys a connection carries beyond the canonical fields.
const (
	AttrCID          = "cid"
	AttrLanguage     = "language"
	AttrLibVersion   = "libVersion"
	AttrIdleTime     = "idle"
	AttrLastActivity = "lastActivity"
	AttrPendingBytes = "pendingBytes"
	AttrInMsgs       = "inMsgs"
	AttrOutMsgs      = "outMsgs"
	AttrRTT          = "rtt"
	AttrSubjectList  = "subjectList"
	AttrAccount      = "account"
	AttrKind         = "kind"
)

// maxConnections bounds one listing. The endpoint pages, and a server with
// more connections than this is one where a list was never the useful view.
const maxConnections = 1024

// ListClientConnections enumerates the sockets open against the cluster.
//
// Through the system account where there is one, for the same reason the
// server listing is: /connz answers for the one server it belongs to, and a
// client connected to any other server in the cluster simply does not appear.
// On a three-server cluster that is two thirds of the connections missing,
// with nothing on the page to say so.
//
// The namespace argument is the account to narrow to. Empty means every
// account the credentials can see, which for the monitoring endpoint is all of
// them and for the system account is likewise.
func (c *Conn) ListClientConnections(ctx context.Context, namespace string) ([]*model.ClientConnection, error) {
	if err := c.requireClusterSource(); err != nil {
		return nil, err
	}

	query := url.Values{
		// Subscriptions are the useful part of a connection here: NATS has no
		// consumer object outside JetStream, so what a client is subscribed to
		// is the only answer to "what is this connection doing".
		"subs":  {"true"},
		"limit": {strconv.Itoa(maxConnections)},
	}
	if account := strings.TrimSpace(namespace); account != "" {
		query.Set("acc", account)
	}

	if c.system != nil {
		replies, err := c.system.pingWithBody(ctx, endpointConnz, connzRequest(namespace), 0)
		if err == nil && len(replies) > 0 {
			return connectionsFromReplies(replies)
		}
		if c.monitor == nil {
			return nil, err
		}
	}

	var response connzResponse
	if err := c.monitor.get(ctx, pathConnz, query, &response); err != nil {
		return nil, err
	}
	return connectionsOf(response.Server.Name, SourceMonitor, response.Connections), nil
}

// ListClientChannels is empty, and that is a fact rather than a gap.
//
// A NATS connection has no second layer inside it. AMQP multiplexes channels
// over one socket and RabbitMQ's pages are built around that; here a
// connection is one client, and its subscriptions are not channels - they have
// no state, no flow control and no independent lifetime.
func (c *Conn) ListClientChannels(ctx context.Context, namespace string) ([]*model.ClientChannel, error) {
	return []*model.ClientChannel{}, nil
}

// CloseClientConnection disconnects one client.
//
// The system account only. There is no way to close a connection through the
// monitoring endpoint - it is read-only by design - so an endpoint with only
// that reports the capability as degraded rather than offering a button that
// cannot work.
func (c *Conn) CloseClientConnection(ctx context.Context, name, reason string) error {
	if c.system == nil {
		return &driverUnsupported{reason: c.systemReasonOr()}
	}
	serverName, cid, err := splitConnectionKey(name)
	if err != nil {
		return err
	}
	return c.system.kick(ctx, serverName, cid)
}

// CloseUserConnections closes every connection one identity holds.
//
// NATS has no request for this: KICK takes a client id on a named server, and
// there is no "disconnect this user" call anywhere in $SYS. So this lists,
// filters and kicks each - which is what an operator would do by hand, and is
// worth doing here because an application with several instances is otherwise
// evicted one row at a time.
func (c *Conn) CloseUserConnections(ctx context.Context, username, reason string) error {
	if c.system == nil {
		return &driverUnsupported{reason: c.systemReasonOr()}
	}
	connections, err := c.ListClientConnections(ctx, "")
	if err != nil {
		return err
	}

	var closed int
	for _, connection := range connections {
		if connection.User != username {
			continue
		}
		serverName, cid, err := splitConnectionKey(connection.Name)
		if err != nil {
			continue
		}
		if err := c.system.kick(ctx, serverName, cid); err != nil {
			return err
		}
		closed++
	}
	if closed == 0 {
		return fmt.Errorf("no connection is open as %q", username)
	}
	return nil
}

func (c *Conn) systemReasonOr() string {
	if c.tiers.systemReason != "" {
		return c.tiers.systemReason
	}
	return systemAbsent
}

/*
 * A connection is addressed by the server holding it and its client id.
 *
 * Neither half is enough on its own: a client id counts within one server, so
 * two servers in a cluster will each have a client 7. The canonical model has
 * one Name field and it is what a close request names, so the two are joined -
 * and split again on the way back out.
 */
func connectionKey(server string, cid uint64) string {
	return fmt.Sprintf("%s/%d", server, cid)
}

func splitConnectionKey(name string) (string, uint64, error) {
	server, raw, found := strings.Cut(name, "/")
	if !found {
		return "", 0, fmt.Errorf(
			"%q does not name a connection; a NATS connection is a server and a client id", name)
	}
	cid, err := strconv.ParseUint(raw, 10, 64)
	if err != nil {
		return "", 0, fmt.Errorf("%q does not end in a client id", name)
	}
	return server, cid, nil
}

// connzResponse is the subset of /connz this driver reads.
type connzResponse struct {
	Server struct {
		Name string `json:"server_name"`
		ID   string `json:"server_id"`
	} `json:"server"`
	Now         string            `json:"now"`
	Total       int               `json:"total"`
	Connections []connzConnection `json:"connections"`
}

type connzConnection struct {
	CID        uint64   `json:"cid"`
	Kind       string   `json:"kind"`
	Type       string   `json:"type"`
	IP         string   `json:"ip"`
	Port       int      `json:"port"`
	Start      string   `json:"start"`
	LastActive string   `json:"last_activity"`
	RTT        string   `json:"rtt"`
	Uptime     string   `json:"uptime"`
	Idle       string   `json:"idle"`
	Pending    int      `json:"pending_bytes"`
	InMsgs     int64    `json:"in_msgs"`
	OutMsgs    int64    `json:"out_msgs"`
	InBytes    int64    `json:"in_bytes"`
	OutBytes   int64    `json:"out_bytes"`
	NumSubs    uint32   `json:"subscriptions"`
	Name       string   `json:"name"`
	Lang       string   `json:"lang"`
	Version    string   `json:"version"`
	TLSVersion string   `json:"tls_version"`
	TLSCipher  string   `json:"tls_cipher_suite"`
	Account    string   `json:"account"`
	User       string   `json:"authorized_user"`
	SubsList   []string `json:"subscriptions_list"`
}

// connzRequest is the body a $SYS CONNZ request takes.
//
// The field is "subscriptions" here and "subs" on the monitoring endpoint's
// query string. They are the same option, and the two spellings are the
// server's rather than a choice made here - sending the query string's name in
// the body is accepted and silently ignored, which is how this was first
// written and why the subject column came back empty.
func connzRequest(namespace string) any {
	request := map[string]any{
		"subscriptions": true,
		"limit":         maxConnections,
	}
	if account := strings.TrimSpace(namespace); account != "" {
		request["acc"] = account
	}
	return request
}

func connectionsFromReplies(replies []systemReply) ([]*model.ClientConnection, error) {
	connections := make([]*model.ClientConnection, 0, 16)
	for _, reply := range replies {
		var response connzResponse
		if err := json.Unmarshal(reply.Data, &response); err != nil {
			return nil, fmt.Errorf("$SYS CONNZ answered something unexpected: %w", err)
		}
		server := response.Server.Name
		if server == "" {
			server = reply.Server.Name
		}
		connections = append(connections, connectionsOf(server, SourceSystem, response.Connections)...)
	}
	sort.Slice(connections, func(i, j int) bool { return connections[i].Name < connections[j].Name })
	return connections, nil
}

func connectionsOf(server, source string, raw []connzConnection) []*model.ClientConnection {
	connections := make([]*model.ClientConnection, 0, len(raw))
	for _, item := range raw {
		connections = append(connections, connectionOf(server, source, item))
	}
	return connections
}

// connectionOf maps one connection onto the canonical model.
//
// The source travels with the row rather than being inferred once for the
// page, because it decides what can be done to that particular connection:
// a row read through the monitoring endpoint cannot be closed at all - that
// endpoint is read-only by design - and a button that was always going to be
// refused is worse than one that is not there.
func connectionOf(server, source string, raw connzConnection) *model.ClientConnection {
	connection := &model.ClientConnection{
		Name:       connectionKey(server, raw.CID),
		ClientName: raw.Name,
		Namespace:  raw.Account,
		User:       raw.User,
		Node:       server,
		PeerHost:   raw.IP,
		PeerPort:   raw.Port,
		// The transport, which on a NATS server can be any of four: a client
		// may arrive over TCP, TLS, WebSocket or MQTT, and treating them alike
		// would misreport all of them.
		Protocol: connectionProtocol(raw),
		State:    "running",
		// A NATS connection has no second layer inside it, so this is not a
		// count that happens to be zero - there is nothing to count.
		Channels:  model.UnknownMetric,
		TLS:       raw.TLSVersion != "",
		Cipher:    raw.TLSCipher,
		RecvBytes: raw.InBytes,
		SendBytes: raw.OutBytes,
		// NATS reports totals rather than rates. A per-second figure would be
		// two samples divided by the time between them.
		RecvByteRate: -1,
		SendByteRate: -1,
		// Heartbeats are the server's ping interval rather than something the
		// two sides negotiate, and /connz does not report it per connection.
		HeartbeatSec: model.UnknownMetric,
		Attributes:   map[string]string{},
	}

	if started, err := time.Parse(time.RFC3339Nano, raw.Start); err == nil {
		connection.ConnectedAtMs = started.UnixMilli()
	}

	set := func(key, value string) {
		if value != "" {
			connection.Attributes[key] = value
		}
	}
	set(AttrSource, source)
	set(AttrCID, strconv.FormatUint(raw.CID, 10))
	set(AttrKind, raw.Kind)
	set(AttrLanguage, raw.Lang)
	set(AttrLibVersion, raw.Version)
	set(AttrIdleTime, raw.Idle)
	set(AttrLastActivity, raw.LastActive)
	set(AttrRTT, raw.RTT)
	set(AttrPendingBytes, strconv.Itoa(raw.Pending))
	set(AttrInMsgs, strconv.FormatInt(raw.InMsgs, 10))
	set(AttrOutMsgs, strconv.FormatInt(raw.OutMsgs, 10))
	set(AttrAccount, raw.Account)
	// What the connection is subscribed to, which is the only answer NATS has
	// to "what is this client doing": there is no consumer object outside
	// JetStream to look it up in.
	set(AttrSubjectList, strings.Join(raw.SubsList, ", "))
	return connection
}

// connectionProtocol names the transport a client arrived over.
func connectionProtocol(raw connzConnection) string {
	switch {
	case raw.Type != "":
		return raw.Type
	case raw.TLSVersion != "":
		return "tls"
	default:
		return "nats"
	}
}

File: frontend/src/mq/nats/clients.ts
/**
 * NATS's view of a canonical client connection.
 *
 * The keys are a contract with internal/driver/nats/clients.go.
 *
 * Two canonical fields are read differently here and both are worth knowing.
 * `name` is the server plus the client id joined by a slash, because neither
 * half addresses a connection on its own - a client id counts within one
 * server, so two servers in a cluster each have a client 7. And `channels` is
 * UnknownMetric rather than zero: a NATS connection has no second layer inside
 * it at all, so there is nothing to count.
 */
import type { ClientConnection } from "@bindings/model/models";

const AttrCID = "cid";
const AttrLanguage = "language";
const AttrLibVersion = "libVersion";
const AttrIdleTime = "idle";
const AttrLastActivity = "lastActivity";
const AttrPendingBytes = "pendingBytes";
const AttrInMsgs = "inMsgs";
const AttrOutMsgs = "outMsgs";
const AttrRTT = "rtt";
const AttrSubjectList = "subjectList";
const AttrAccount = "account";
const AttrKind = "kind";
const AttrSource = "readVia";

function attr(connection: ClientConnection, key: string): string | null {
  const value = connection.attributes?.[key];
  return value == null || value === "" ? null : value;
}

function number(connection: ClientConnection, key: string): number | null {
  const raw = attr(connection, key);
  if (raw == null) return null;
  const value = Number.parseInt(raw, 10);
  return Number.isNaN(value) ? null : value;
}

/** The key a close request names: the server holding it, and its client id. */
export const connectionKey = (connection: ClientConnection): string => connection.name;

/** Which server holds it. Half of the address, not a label. */
export const serverOf = (connection: ClientConnection): string => connection.node;

export const clientId = (connection: ClientConnection): number | null =>
  number(connection, AttrCID);

/**
 * What the application called itself, or null when it said nothing.
 *
 * Most libraries send nothing, which is why the peer address stays the primary
 * identifier and this is shown beside it rather than instead of it.
 */
export function clientName(connection: ClientConnection): string | null {
  const name = connection.clientName?.trim();
  return name == null || name === "" ? null : name;
}

export const peer = (connection: ClientConnection): string =>
  `${connection.peerHost}:${connection.peerPort}`;

export const account = (connection: ClientConnection): string | null =>
  attr(connection, AttrAccount);
export const user = (connection: ClientConnection): string | null =>
  connection.user === "" ? null : connection.user;

/** The transport it arrived over: any of TCP, TLS, WebSocket or MQTT. */
export const transport = (connection: ClientConnection): string => connection.protocol;
export const isEncrypted = (connection: ClientConnection): boolean => connection.tls;
export const cipher = (connection: ClientConnection): string | null =>
  connection.cipher === "" ? null : connection.cipher;

/**
 * What the connection is subscribed to.
 *
 * The only answer NATS has to "what is this client doing": outside JetStream
 * there is no consumer object to look one up in, so the subject list is the
 * whole of it.
 */
export function subjects(connection: ClientConnection): string[] {
  const raw = attr(connection, AttrSubjectList);
  if (raw == null) return [];
  return raw
    .split(",")
    .map((subject) => subject.trim())
    .filter((subject) => subject !== "");
}

export const inMessages = (connection: ClientConnection): number | null =>
  number(connection, AttrInMsgs);
export const outMessages = (connection: ClientConnection): number | null =>
  number(connection, AttrOutMsgs);

/**
 * Bytes the server is holding to send this client.
 *
 * The figure that says a client is falling behind: it is what has been written
 * for them and not yet accepted, and a server disconnects a client whose
 * pending bytes pass its limit.
 */
export const pendingBytes = (connection: ClientConnection): number | null =>
  number(connection, AttrPendingBytes);

export const idleFor = (connection: ClientConnection): string | null =>
  attr(connection, AttrIdleTime);
export const lastActivity = (connection: ClientConnection): string | null =>
  attr(connection, AttrLastActivity);
export const roundTrip = (connection: ClientConnection): string | null =>
  attr(connection, AttrRTT);

/** Which client library, and its version. Often the fastest way to find one. */
export const language = (connection: ClientConnection): string | null =>
  attr(connection, AttrLanguage);
export const libraryVersion = (connection: ClientConnection): string | null =>
  attr(connection, AttrLibVersion);

/** Whether this is an ordinary client or a route, leaf or gateway link. */
export const kind = (connection: ClientConnection): string | null => attr(connection, AttrKind);

export const receivedBytes = (connection: ClientConnection): number => connection.recvBytes;
export const sentBytes = (connection: ClientConnection): number => connection.sendBytes;
export const connectedAtMs = (connection: ClientConnection): number => connection.connectedAtMs;

/**
 * Which tier this row came from, and therefore what can be done to it.
 *
 * A connection read through the monitoring endpoint cannot be closed: that
 * endpoint is read-only by design, so there is no request to make. Carried per
 * row rather than worked out once for the page, because a connection is
 * addressed on the server holding it and the two tiers can answer for
 * different ones.
 */
export const readVia = (connection: ClientConnection): string | null =>
  attr(connection, AttrSource);

export const canBeClosed = (connection: ClientConnection): boolean =>
  readVia(connection) === "system";
